[JSC] Microtask should not care about exceptions

Constellation opened the pull request:
#### 634d22fbdc856a59f02b6527d68a415bf1d7841a
<pre>
[JSC] Microtask should not care about exceptions
<a href="https://bugs.webkit.org/show_bug.cgi?id=245199">https://bugs.webkit.org/show_bug.cgi?id=245199</a>
rdar://99237677

Reviewed by Mark Lam.

Microtask execution should be done without exception, and any exceptions thrown can be cleared when running microtasks.
We clear exceptions extensively for all cases, plus, we early return if termination exception is already thrown.

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):

Canonical link: <a href="https://commits.webkit.org/254546@main">https://commits.webkit.org/254546@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/863596e817f2980e2602dedb1f6a41fcf076018b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89377 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98706 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155012 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93385 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32440 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27942 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81735 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93122 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25765 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76298 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25705 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80567 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68690 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81073 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30204 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/74877 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29937 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15578 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3192 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38537 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/77744 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32093 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34669 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17204 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -79,14 +86,18 @@ void runJSMicrotask(JSGlobalObject* globalObject, MicrotaskIdentifier identifier
-    if (UNLIKELY(globalObject->hasDebugger()))
+    if (UNLIKELY(globalObject->hasDebugger())) {
         globalObject->debugger()->willRunMicrotask(globalObject, identifier);
+        scope.clearException();
```
Shouldn't we also check for `TerminationException` here and return early if needed?  We need to be careful of accidentally clearing the `TerminationException` here.  Maybe make a lambda `clearExceptionIfNotTermination` and use that?  The lambda can return `true` for success, and `false` for failure to clear.  If it returns false (i.e. we've got a termination), return early.

Also, I'm not sure what to do about the inspector given that we called `willRunMicrotask` earlier.  Maybe ask the inspector folks about this.  Alternatively, we can also use a `DeferTerminationForAWhileScope` around the inspector calls and ensure that we don't have to deal with termination conditions.  Yes, I think this is the better thing to do.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -66,9 +66,16 @@ Ref<Microtask> createJSMicrotask(VM& vm, JSValue job, JSValue argument0, JSValue
+    scope.clearException();
+
     auto handlerCallData = JSC::getCallData(job);
+    scope.clearException();
```
Shouldn't we also check for TerminationException here and return early if needed?

shvaikalesh replied:
Don't we always supply the `job` itself here? Like it can never come from userland code. Maybe we can `ASSERT` that there is no exception after `getCallData`?

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -79,14 +86,18 @@ void runJSMicrotask(JSGlobalObject* globalObject, MicrotaskIdentifier identifier
 
-    if (UNLIKELY(globalObject->hasDebugger()))
+    if (UNLIKELY(globalObject->hasDebugger())) {
         globalObject->debugger()->didRunMicrotask(globalObject, identifier);
+        scope.clearException();
```
Let's put this in a `DeferTerminationForAWhileScope`.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -79,14 +86,18 @@ void runJSMicrotask(JSGlobalObject* globalObject, MicrotaskIdentifier identifier
+    }
 
     profiledCall(globalObject, ProfilingReason::Microtask, job, handlerCallData, jsUndefined(), handlerArguments);
     scope.clearException();
```
Let's call the `clearExceptionIfNotTermination` lambda (here and every other place in this function) instead.

Constellation replied:
Done.

MenloDorian reviewed (approved):
r=me with fixes.

Constellation reviewed (commented):
Thanks!

Constellation committed:
[JSC] Microtask should not care about exceptions
https://bugs.webkit.org/show_bug.cgi?id=245199
rdar://99237677

Reviewed by Mark Lam.

Microtask execution should be done without exception, and any exceptions thrown can be cleared when running microtasks.
We clear exceptions extensively for all cases, plus, we early return if termination exception is already thrown.

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):

Canonical link: https://commits.webkit.org/254546@main
